primitives/src/unique_vec.rs

```diff
@@ -0,0 +1,334 @@
+
+/// Creates a [`UniqueVec`](unique_vec::UniqueVec) from a list of values.
+///
+/// Works like [`vec!`] macro, but does not accept syntax for repeated values
+/// and might return [`Result`].
+#[macro_export]
+macro_rules! unique_vec {
+    () => {
+        $crate::unique_vec::UniqueVec::new()
+    };
+    ($($x:expr),+ $(,)?) => {
+        $crate::unique_vec::UniqueVec::try_from(vec![$($x),+])
+    };
+}
+
+/// Duplicated value at index {index}
+#[cfg_attr(feature = "std", derive(thiserror::Error))]
+#[derive(Debug, displaydoc::Display, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
+pub struct DuplicationError {
+    index: usize,
+}
+
+/// Wrapper type for [`Vec`] which ensures that all elements are unique.
+#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord, Serialize, Encode, IntoSchema)]
+pub struct UniqueVec<T: PartialEq>(Vec<T>);
+
+impl<T: PartialEq> Default for UniqueVec<T> {
+    fn default() -> Self {
+        Self(Vec::default())
+    }
+}
+
+impl<T: PartialEq> UniqueVec<T> {
+    /// Create new [`UniqueVec`].
+    #[must_use]
+    pub const fn new() -> Self {
+        Self(Vec::new())
+    }
+
+    /// Try construct [`UniqueVec`] from iterator.
+    ///
+    /// # Errors
+    ///
+    /// - If there are duplicates in the sequence.
+    pub fn try_from_iter<I>(iter: I) -> Result<Self, DuplicationError>
+    where
+        I: IntoIterator<Item = T>,
+    {
+        let mut v = UniqueVec::new();
+        v.try_extend(iter)?;
+        Ok(v)
+    }
+
+    /// Try extend [`UniqueVec`] from iterator.
+    /// If an error occurs, all previously added elements will remain in the [`UniqueVec`].
+    ///
+    /// # Errors
+    ///
+    /// - If there are duplicates in the sequence.
+    pub fn try_extend<I>(&mut self, iter: I) -> Result<(), DuplicationError>
+    where
+        I: IntoIterator<Item = T>,
+    {
+        for (i, value) in iter.into_iter().enumerate() {
+            if !self.push(value) {
+                return Err(DuplicationError { index: i });
+            }
+        }
+        Ok(())
+    }
+
+    /// Push `value` to [`UniqueVec`] if it is not already present.
+    ///
+    /// Returns `true` if value was pushed and `false` if not.
+    pub fn push(&mut self, value: T) -> bool {
+        if self.contains(&value) {
+            false
+        } else {
+            self.0.push(value);
+            true
+        }
+    }
+
+    /// Removes the element at the given `index` and returns it.
+    ///
+    /// # Panics
+    ///
+    /// Panics if the `index` is out of bounds.
+    pub fn remove(&mut self, index: usize) -> T {
+        self.0.remove(index)
+    }
+
+    /// Clears the [`UniqueVec`], removing all values.
+    pub fn clear(&mut self) {
+        self.0.clear()
+    }
+
+    /// Returns reference to an internal mutable slice.
+    ///
+    /// # Warning
+    ///
+    /// Caller is responsible for ensuring that elements in slice will remain unique.
+    pub fn as_mut_unchecked(&mut self) -> &mut [T] {
```

Comment:
having this method can break the invariant of uniqueness and be the source of potential bugs. I would remove it

Reply:
I think that slapping `_unchecked` is only for constructors. I don't think you can apply it to getters. Can you point me to some example in `std`

Reply:
it's not that the method is `unchecked` and if you maintain invariants at that place all will be good. The problem is that consumers of the slice you return `&mut [T]` have no way of knowing they can break the code by mutating it

Reply:
you should implement mutating methods directly on the `UniqueVec` without exposing mutable slice directly

Reply:
or you can drain `UniqueVec` into a `Vec` do the operation and collect it back into `UniqueVec`

Reply:
> or you can drain UniqueVec into a Vec do the operation and collect it back into UniqueVec. This sounds even better to me

Nice, that should fix the original problem!